### DISCUSSION TITLE
White Screen with ST7735S and ESP8266

### DISCUSSION THREAD

**Discussion by JooJooBee666** [Bodmer/TFT_eSPI#2814] [Compatible displays and setup files]
No matter what configuration and pins I set, I cannot get the screen to initialize.  If I leave the pins the same, then plug the screen in to an ESP32 using the same sketch, the screen works fine.  This is clearly an issue with ESP8266 only.

More Details:

IDE: Arduino 2.11 and PLatformIO with VSCode
TFT_eSPI ver: 2.5.0(latest)
Processor: ESP8266
Driver: ST7735
Interface: SPI (of course!)
Current User_Setup.h (that works on ESP32):

```
//#define USER_SETUP_ID 7

#define ST7735_DRIVER


#define TFT_WIDTH  128
#define TFT_HEIGHT 160

#define ST7735_BLACKTAB

// Generic ESP8266 setup
#define TFT_MOSI  12
#define TFT_SCLK  4
#define TFT_CS    2 // Not defined here, chip select is managed by sketch
#define TFT_DC    0
#define TFT_RST   14 // Connect reset to ensure display initialises


#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
//#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
//#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:.
//#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts


#define SMOOTH_FONT


// #define SPI_FREQUENCY  20000000
#define SPI_FREQUENCY  27000000
```

I just used the example **Colour_Test** sketch to test this.   I also ran the **Read_User_Setup** to verify that the pins were being read correctly.  To further verify, I made another sketch with a simple PIN HIGH/LOW and hooked up an LED to each PIN I'm trying to use and verified they all respond properly by blinking the LED.  I was even able to use the **Read_ID_bitbash** and it tells me I have an **ST7735S**.  So it seems like my pins are correct but just stuck on a white screen no matter what I do.  I'm stuck, out of options and don't know enough about how the screens work to look deeper in this code to figure out what's happening yet.  However, I'm hoping that the fact that this works fine on ESP32 but not ESP8266 means it just some errant **#ifdef** somewhere breaking this.

The last bit of info I can give is that the screen works just fine using the **Adafruit_ST7735** driver library.  So it can and does work on the ESP8266.

There isn't much output I can get.  If you absolutely want to see the **Read_ID_bitbash** output or **Read_User_Setup** output, I can paste it but I just consider it redundant at this point given the config above is what is being used, read and working on a different chip.  However, if you need any more information, please let me know.

**Comment by Bodmer** [comment]
You must use the dedicated hardware SPI pins on an ESP8266. The Read_ID_bitbash is able to use any pins as it uses bit bashing, not the SPI hardware.

```
// Typical setup for ESP8266 NodeMCU ESP-12 is :
//
// Display SDO/MISO  to NodeMCU pin D6 (or leave disconnected if not reading TFT)
// Display LED       to NodeMCU pin VIN (or 5V, see below)
// Display SCK       to NodeMCU pin D5
// Display SDI/MOSI  to NodeMCU pin D7
// Display DC (RS/AO)to NodeMCU pin D3
// Display RESET     to NodeMCU pin D4 (or RST, see below)
// Display CS        to NodeMCU pin D8 (or GND, see below)
// Display GND       to NodeMCU pin GND (0V)
// Display VCC       to NodeMCU 5V or 3.3V
```

If you have a different ESP8266 board then the mapping of the Dx pins on a NodeMCU board to the GPIO is as follows:

// These are the pins for ESP8266 boards
//      Name   GPIO    NodeMCU      Function
#define PIN_D0  16  // GPIO16       WAKE
#define PIN_D1   5  // GPIO5        User purpose
#define PIN_D2   4  // GPIO4        User purpose
#define PIN_D3   0  // GPIO0        Low on boot means enter FLASH mode
#define PIN_D4   2  // GPIO2        TXD1 (must be high on boot to go to UART0 FLASH mode)
#define PIN_D5  14  // GPIO14       HSCLK
#define PIN_D6  12  // GPIO12       HMISO
#define PIN_D7  13  // GPIO13       HMOSI  RXD2
#define PIN_D8  15  // GPIO15       HCS    TXD0 (must be low on boot to enter UART0 FLASH mode)
#define PIN_D9   3  //              RXD0
#define PIN_D10  1  //              TXD0

#define PIN_MOSI 8  // SD1          FLASH and overlap mode
#define PIN_MISO 7  // SD0
#define PIN_SCLK 6  // CLK
#define PIN_HWCS 0  // D3

#define PIN_D11  9  // SD2
#define PIN_D12 10  // SD4

**Reply by JooJooBee666** [reply]
Sorry, I still don't know what I need to do here.  My board has those listing...but quite a few do not align with the pins defined in User_Setup_select.h/with that you pasted here.  Here's how my board is labeled (at least regarding the SPI pins):

```
D13/SCK - GPIO 14
D12/MISO - GPIO 12
D11/MOSI - GPIO 13/RX0
D10/SS - GPIO 15/TX0
TX1/D9 - GPIO 2
D8 - GPIO 0
D11/MOSI/D7 - GPIO 13
D12/MISO/D6 - GPIO 12
D13/SCK/D5 - GPIO 15
D14/SDA/D4 - GPIO 4
D15/SCL/D3 - GPIO 5
```
Can you tell me exactly what I'm supposed to do with this?  None of it quite aligns with anything in the documentation, leaving me to try all kinds of combinations to no avail.  Here's what I currently tried but it did not work:
```
#define TFT_MOSI  PIN_D7
#define TFT_SCLK  PIN_D5
#define TFT_CS    PIN_D8
#define TFT_DC    PIN_D3
#define TFT_RST   PIN_D4
```
I also changed the wiring so that they actually matched the REAL GPIO pins after trying it without, but that failed too. I should also note that I also tried assigning the GPIO directly in the setup as they are labeled on the board to line up as well, no dice.

Also, what's the reason I cannot choose the pins I want?  I can do that with the Adafruit_ST7735 driver.

**Reply by JooJooBee666** [reply]
@Bodmer I also just realized you closed the issue and put this on the discussion board.  My issue is not resolved.  Can we reopen this as well?  I'm don't think the given answer helps others that are having the same issue as me.

Thanks. :clinking_glasses:

**Reply by JooJooBee666** [reply]
Well, one last reply.  I decided to take one last crack at it and FINALLY got it working; using yet AGAIN another pin-out 
I hadn't tried yet.  I really think the documentation for this needs some clarifications.  This is just not intuitive at all.  Why is this the only pinout working with TFT_eSPI with my board?

```
#define TFT_MOSI  PIN_D7 // (GPIO 13)
#define TFT_SCLK  PIN_D5 // (GPIO 15)
#define TFT_CS    PIN_D2 // (GPIO 4)
#define TFT_DC    PIN_D4 // (GPIO 2)
#define TFT_RST   PIN_D3 // (GPIO 0)
```

**Comment by JooJooBee666** [comment]
@Bodmer I'm just following up here.  Can you explain why the ESP8266 is so difficult to get going with TFT_eSPI and why only certain pins work?  I'm just trying to understand a bit better here.

**Comment by ustaslastik** [comment]
I have a Bruce ESP32-S3 board and a 1.8-inch TFT display with ST7735 driver.

Display info:

* Driver: ST7735
* Resolution: 128x160
* Touch: No
* SKU: MAR1801

Pins on the display:
VCC
GND
GND
NC
NC
NC
CLK
SDA
RS
RST
CS

SD card pins:
SD_CS
SD_MOSI
SD_SCK
SD_MISO

I am trying to connect it to ESP32-S3, but the screen only shows a white display.

Does anyone know the correct wiring and TFT_eSPI configuration for this display?

I would be very grateful if someone could help me. Thank you in advance.
